### PR TITLE
docs: Add central contributing doc

### LINKED ISSUE #5: Create a central contributing document
We need to create a `CONTRIBUTING.md` document in this repository whose content can be applied to all the other Kata (code) repositories.

Once done, we can create a minimal `CONTRIBUTING.md` in each other repo that says something like:

> For details on how to contribute to the Kata Containers project, please see the main [contributing document](https://github.com/kata-containers/community/blob/master/CONTRIBUTING.md).

References:

- https://github.com/kata-containers/proxy/pull/14
- https://github.com/kata-containers/proxy/pull/15

### PR DESCRIPTION
**Pull request by jodh-intel**
This doc will be referenced by all the other Kata Containers
repositories.

Fixes #5.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

### CODE REVIEW

**Comment by jodh-intel** [comment]
Technically, #8 should land before this given that this doc references the `CODE_OF_CONDUCT.md`.

Note that the `CONTRIBUTING.md` doc on this PR is heavily based on:

- https://github.com/clearcontainers/runtime/blob/master/CONTRIBUTING.md

In fact, I'd recommend `diff`-ing the two to see "what changed" :smile:

/cc @sameo, @grahamwhaley, @bergwolf.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+used by the Linux kernel project.
+
+## Pull requests
+
+All the repositories accept contributions by a github Pull request (PR).
+
+### Before starting work on a PR
+
+We welcome all contributions but to minimise the chance of multiple people
+working on the same feature or bug fixe (yes, it has happened), we recommend
```
Comment (grahamwhaley) [inline CONTRIBUTING.md]: s/fixe/fixes/

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+comments section should be re-worked into your patch set, as the ultimate place
+where your patches are documented is in the git log, and not in the github
+comments section.
+
+For more information on github "force push" workflows see
+[here](http://blog.adamspiers.org/2015/03/24/why-and-how-to-correctly-amend-github-pull-requests/).
+
+It is perfectly fine for your PR to contain more than one patch - use
+as many patches as you need to implement the Request (see the previously
+mentioned "small patch" thoughts). Each PR should only cover one
```
Comment (grahamwhaley) [inline CONTRIBUTING.md]: is there a previous 'small patch' thought bit, in this doc?

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+## Patch format
+
+Beside the `Signed-off-by` footer, we expect each patch to comply with the
+following format:
+
+```
+Subsystem: Change summary (no longer than 75 characters)
+
+More detailed explanation of your changes: Why and how.
+Wrap it to 72 characters.
```
Comment (sameo) [inline CONTRIBUTING.md]: Could we please remove that particular check?

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+
+  For example, it might be that an important bug fix or new feature is
+  required. It might even be that a particular commit is needed. If so, please
+  state that.
+
+- The range of commits being added for these third-party packages.
+
+  It is possible that re-vendoring a particular package will also result in updates to
+  other dependent packages. However, it is important to include the commit range
+  (even if it is a big) for the primary package(s) the re-vendor PR has been
```
Comment (grahamwhaley) [inline CONTRIBUTING.md]: s/is a big/is big/ ?

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+  raised for.
+
+  These details allow for easier problem-determination in the future should
```
Comment (grahamwhaley) [inline CONTRIBUTING.md]: s/problem-determination/problem determination/

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+
+Correct formatting of the PR patches is verified using the
+[checkcommits](https://github.com/kata-containers/tests/tree/master/cmd/checkcommits)
+tool.
+
+Note, that the body of the message should not just be a continuation of the
+subject line, and is not used to extend the subject line beyond its length
+limit. They should stand alone as complete sentence and paragraphs.
+
+It is recommended that each of your patches fixes one thing. Smaller patches
```
Comment (grahamwhaley) [inline CONTRIBUTING.md]: ah, and here is the 'small patches' bit - so above we probably need to say 'the below' or similar when referencing it - or just 'elsewhere in this document' etc.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+
+The Kata Containers community can be reached through its dedicated IRC
+channels and mailing lists:
+
+* IRC:
+  * Development discussions: `#kata-dev @ freenode.net`.
+  * General discussions: `#kata-general @ freenode.net`.
+
+* [Mailing lists](http://lists.katacontainers.io/).
+
```
Comment (grahamwhaley) [inline CONTRIBUTING.md]: should we add a slack ref in here as well?

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+If it is unclear which repository you should be raising your query against,
+first, try getting in [contact](#contact) with us. But if in doubt, raise the
+issue [here](https://github.com/kata-containers/community/issues/new) and we
+will help you to handle the query by routing it to the correct area for
+resolution.
+
+## Closing issues
+
+The preferred way to close issues is by adding the `Fixes` keyword to your
+commit message. Our tooling requires a `Fixes` line in at least one commit
```
Comment (grahamwhaley) [inline CONTRIBUTING.md]: s/your commit message/the relevant commit message in a PR/

**Comment by grahamwhaley** [comment]
@sameo - should this doc get some form of committee review before it lands?

**Comment by sameo** [comment]
@grahamwhaley Unless there are some strong disagreement within the maintainers, the committee should not be involved. If we all agree about that document, we can go ahead and merge it.

File: CONTRIBUTING.md (hunk 3 above)
Comment (jodh-intel) [reply CONTRIBUTING.md]: Oh good catch (see https://github.com/kata-containers/agent/pull/77#discussion_r158183877). Removing...

File: CONTRIBUTING.md (hunk 1 above)
Comment (jodh-intel) [reply CONTRIBUTING.md]: Fixing...

File: CONTRIBUTING.md (hunk 2 above)
Comment (jodh-intel) [reply CONTRIBUTING.md]: Another good catch - this is present in the CC runtime's version of this doc too. Removing...

File: CONTRIBUTING.md (hunk 4 above)
Comment (jodh-intel) [reply CONTRIBUTING.md]: Fixing...

File: CONTRIBUTING.md (hunk 5 above)
Comment (jodh-intel) [reply CONTRIBUTING.md]: Yep. It seems that hyphens are becoming deprecated everywhere (sob), so removing...

File: CONTRIBUTING.md (hunk 7 above)
Comment (jodh-intel) [reply CONTRIBUTING.md]: Good idea. Adding...

File: CONTRIBUTING.md (hunk 8 above)
Comment (jodh-intel) [reply CONTRIBUTING.md]: That paragraph contains some duplication, so I'll rewrite it...

**Comment by jodh-intel** [comment]
Branch updated.

@gnawux, @bergwolf, @klynnrif - could you take a look please?

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+at least one approval from one member of each list for a PR to be merged.
+
+Since approvers are also reviewers, they technically can approve a PR without
+getting another reviewer's approval. However, it is their due diligence to
+rely on reviewers and should use their approval power only in very specific
+cases.
+
+## Issue tracking
+
+To report a bug that is not already documented, please open a github issue for
```
Comment (klynnrif) [inline CONTRIBUTING.md]: ...please open a GitHub issue for

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+the repository in question.
+
+If it is unclear which repository you should be raising your query against,
```
Comment (klynnrif) [inline CONTRIBUTING.md]: If it is unclear which repository to raise your query against, first try to get in [contact](#contact) with us. If in doubt, raise the

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+first, try getting in [contact](#contact) with us. But if in doubt, raise the
+issue [here](https://github.com/kata-containers/community/issues/new) and we
+will help you to handle the query by routing it to the correct area for
+resolution.
+
+## Closing issues
+
+Our tooling requires a `Fixes` comment to be added to atleast one commit in
```
Comment (klynnrif) [inline CONTRIBUTING.md]: Lines 254-256 suggested rewrite: Our tooling requires adding a `Fixes `comment to at least one commit in the PR, which triggers GitHub to automatically close the issue once the PR is merged:

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+
+The token and pid data will be hold by the new Process structure and
+they are related to a container.
+
+Fixes #123
+
+Signed-off-by: Sebastien Boeuf <sebastien.boeuf@intel.com>
+```
+
+Github will then automatically close that issue when parsing the
```
Comment (klynnrif) [inline CONTRIBUTING.md]: Lines 269-270 suggested rewrite: The issue is automatically closed by GitHub when the [commit message](https://help.github.com/articles/closing-issues-via-commit-messages/) is parsed.

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+Correct formatting of the PR patches is verified using the
+[checkcommits](https://github.com/kata-containers/tests/tree/master/cmd/checkcommits)
+tool.
+
+Note, that the body of the message should not just be a continuation of the
+subject line, and is not used to extend the subject line beyond its length
+limit. They should stand alone as complete sentence and paragraphs.
+
+It is recommended that each of your patches fixes one thing. Smaller patches
+are easier to review, and are thus more likely to be accepted and merged, and
```
[the remark this answers is not in the document]
Comment (jodh-intel) [reply CONTRIBUTING.md]: Changed with a small correction:

> are easier to review, more likely accepted and merged, and problems are more
> **to be** likely identified during review.

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+  (even if it is big) for the primary package(s) the re-vendor PR has been
+  raised for.
+
+  These details allow for easier problem determination in the future should
+  the re-vendor PR introduce bugs or behavioural changes.
+
+  Generate the list of new commits added to the primary re-vendored
+  package by comparing the previous and latest commits for the package being re-vendored.
+
+  For example, if a new version of `libcontainer` (part of the `runc`)
```
[the remark this answers is not in the document]
Comment (jodh-intel) [reply CONTRIBUTING.md]: Changed to correct bracketing:

> The following example lists the steps you should take if a new version of
> `libcontainer` (part of the `runc` repository) is required:

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -0,0 +1,270 @@
...
+
+## Certificate of Origin
+
+In order to get a clear contribution chain of trust we use the [signed-off-by
+language](https://ltsi.linuxfoundation.org/software/signed-off-process/)
+used by the Linux kernel project.
+
+## Pull requests
+
+All the repositories accept contributions by a github Pull request (PR).
```
[the remark this answers is not in the document]
Comment (jodh-intel) [reply CONTRIBUTING.md]: All instances updated to "GitHub".

File: CONTRIBUTING.md [hunk 16]
```diff
@@ -0,0 +1,270 @@
...
+limit. They should stand alone as complete sentence and paragraphs.
+
+It is recommended that each of your patches fixes one thing. Smaller patches
+are easier to review, and are thus more likely to be accepted and merged, and
+problems are more likely to be picked up during review.
+
+## Reviews
+
+Before your PRs are merged into the main code base, they will be
+reviewed. Anybody can review any PR and leave feedback (in fact, it
```
[the remark this answers is not in the document]
Comment (jodh-intel) [reply CONTRIBUTING.md]: This does change the meaning of the sentence slightly. But it's implicit that anybody has the power to review any PR I think, so applying...

File: CONTRIBUTING.md [hunk 17]
```diff
@@ -0,0 +1,270 @@
...
+  * General discussions: `#kata-general @ freenode.net`.
+
+* [Slack channels](https://katacontainers.slack.com/) ([invite](http://bit.ly/KataSlack)).
+
+* [Mailing lists](http://lists.katacontainers.io/).
+
+### Project maintainers
+
+The Kata Containers project maintainers will be the ones accepting or
+rejecting any PR. They are listed in the `OWNERS` files, and there can be one
```
[the remark this answers is not in the document]
Comment (jodh-intel) [reply CONTRIBUTING.md]: Changed to the following to retain the original intent:

>  They are listed in the `OWNERS` files. There can be one
> `OWNERS` file per directory.

File: CONTRIBUTING.md [hunk 18]
```diff
@@ -0,0 +1,270 @@
...
+language](https://ltsi.linuxfoundation.org/software/signed-off-process/)
+used by the Linux kernel project.
+
+## Pull requests
+
+All the repositories accept contributions by a github Pull request (PR).
+
+### Before starting work on a PR
+
+We welcome all contributions but to minimise the chance of multiple people
```
[the remark this answers is not in the document]
Comment (jodh-intel) [reply CONTRIBUTING.md]: Yes, I think we need to specify this in a document somewhere (and probably permit both). For the time being I've switched to the American spelling.

Ping @grahamwhaley - pleaze enzure you read this! Arf! :smile:

**Comment by jodh-intel** [comment]
Phew! Thanks for the thorough review @klynnrif. Branch updated.

**Comment by klynnrif** [comment]
Anytime @jodh-intel! :)

**Comment by bergwolf** [comment]
lgtm!

**Pull request merged**